core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp

```diff
@@ -92,12 +103,24 @@ class Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>,
     sycl::queue& q = *instance.m_queue;
 
     auto parallel_for_event = q.submit([&](sycl::handler& cgh) {
-      FunctorWrapperRangePolicyParallelFor<Functor, Policy> f{policy.begin(),
-                                                              functor};
-      sycl::range<1> range(policy.end() - policy.begin());
       cgh.depends_on(memcpy_event);
-      cgh.parallel_for<FunctorWrapperRangePolicyParallelFor<Functor, Policy>>(
-          range, f);
+      if constexpr (Policy::launch_bounds::maxTperB == 0) {
+        FunctorWrapperRangePolicyParallelFor<Functor, Policy> f{policy.begin(),
+                                                                functor};
+        sycl::range<1> range(policy.end() - policy.begin());
+        cgh.parallel_for<FunctorWrapperRangePolicyParallelFor<Functor, Policy>>(
+            range, f);
+      } else {
+        const auto actual_range = policy.end() - policy.begin();
+        const auto wgroup_size  = Policy::launch_bounds::maxTperB;
+        const auto launch_range =
+            (actual_range + wgroup_size - 1) / wgroup_size * wgroup_size;
+        FunctorWrapperRangePolicyParallelFor<Functor, Policy> f{
+            policy.begin(), functor, actual_range};
+        sycl::nd_range<1> range(launch_range, Policy::launch_bounds::maxTperB);
```

Comment:
To clarify -- the purpose of this code is to make sure `launch_range` is a multiple of the workgroup size? Might be worth commenting this so it is clear what's going on here.

Reply:
Looks good to me @nmm0

Reply:
Add a comment to capture https://github.com/kokkos/kokkos/pull/4875/files?diff=unified&w=1#r827960591

Reply:
[](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)[df1a78b](https://github.com/kokkos/kokkos/pull/4875/commits/df1a78be3e7bb00be6fe7d14533f54a4952974a0)